File: apps/server/shared/practice/schema.py
"""练习相关 Schema 定义"""

from typing import Any, Optional

from pydantic import BaseModel, Field


class SubmitAnswerSchema(BaseModel):
    """提交答案请求"""

    session_id: str = Field(..., description="练习会话 ID (UUID v4)")
    question_id: str = Field(..., description="题目 ID")
    answer: Any = Field(..., description="学生答案（单值、列表或 JSON 字典）")
    time_spent: int = Field(..., description="答题耗时（秒）")
    audio_url: Optional[str] = Field(default=None, description="音频答案URL")


class AnswerAnalysisSchema(BaseModel):
    """答题分析响应（AI 返回格式）"""

    is_correct: bool = Field(default=False, description="是否正确")
    analysis: str = Field(default="", description="错题分析")


class CorrectAnswerSchema(BaseModel):
    """结构化正确答案

    根据题目的 interaction_type 返回对应格式的正确答案，
    前端可根据 type 字段选择渲染方式。
    """

    type: str = Field(..., description="与 interaction_type 对应")
    value: Optional[Any] = Field(default=None, description="单值答案（单选题等）")
    values: Optional[list] = Field(default=None, description="多值答案（多选题等）")
    options: Optional[list] = Field(
        default=None,
        description="选项详情（含文本），格式: [{'id': 'A', 'text': '选项内容'}]",
    )
    sub_answers: Optional[list] = Field(
        default=None,
        description="复合题子答案，格式: [{'sub_id': '1', 'is_correct': bool, 'value': 'A'}]",
    )


class RubricCriterionSchema(BaseModel):
    """评分标准条目"""

    points: int = Field(..., description="分值")
    description: str = Field(..., description="评分描述")


class RubricAnswerSchema(BaseModel):
    """评分量表"""

    criteria: list[RubricCriterionSchema] = Field(..., description="评分标准列表")
    total_points: int = Field(..., description="总分")


class AnswerFeedbackSchema(BaseModel):
    """答题反馈

    错题时返回的完整反馈信息，包含：
    - 结构化正确答案（供前端渲染）
    - 题目自带解析
    - AI 针对性分析
    """

    correct_answer: CorrectAnswerSchema = Field(..., description="结构化正确答案")
    explanation: Optional[str] = Field(default=None, description="题目自带解析")
    analysis: Optional[str] = Field(default=None, description="AI 针对性分析")


class AIRubricEvaluationSchema(BaseModel):
    """AI Rubric 评分结果（LLM 返回格式）

    用于主观题的 AI 评分，返回总分和反馈。
    """

    score: float = Field(..., description="得分（0 到 full_score 之间）")
    is_pass: bool = Field(..., description="是否及格（达到 60% 即为及格）")
    feedback: str = Field(..., description="评分反馈，说明得分/扣分原因")


__all__ = [
    "SubmitAnswerSchema",
    "AnswerAnalysisSchema",
    "CorrectAnswerSchema",
    "AnswerFeedbackSchema",
    "AIRubricEvaluationSchema",
]

File: apps/server-go/internal/ai/schema.go
package ai

import "time"

type QuestionGenerator interface {
	GenerateQuestions(ctx Context, req GenerateQuestionRequest) ([]GeneratedQuestion, error)
}

type AnswerEvaluator interface {
	EvaluateAnswer(ctx Context, req EvaluateAnswerRequest) (EvaluateAnswerResult, error)
}

type ReportGenerator interface {
	GenerateReport(ctx Context, req GenerateReportRequest) (PracticeReportDraft, error)
}

type Context interface {
	Done() <-chan struct{}
	Err() error
}

type GenerateQuestionRequest struct {
	QuestionTypeCode string         `json:"question_type_code"`
	Subject          string         `json:"subject"`
	Grade            int            `json:"grade"`
	Count            int            `json:"count"`
	AbilityCode      string         `json:"ability_code,omitempty"`
	UnitID           int64          `json:"unit_id,omitempty"`
	UnitContent      string         `json:"unit_content,omitempty"`
	Prompt           string         `json:"prompt,omitempty"`
	Params           map[string]any `json:"params,omitempty"`
}

type GeneratedQuestion struct {
	ID               string          `json:"id"`
	QuestionTypeCode string          `json:"question_type_code"`
	Subject          string          `json:"subject"`
	Grade            int             `json:"grade"`
	Content          QuestionContent `json:"content"`
	Answer           QuestionAnswer  `json:"answer"`
	Difficulty       string          `json:"difficulty,omitempty"`
}

type QuestionContent struct {
	Stem         string            `json:"stem"`
	Resource     *Resource         `json:"resource,omitempty"`
	Options      []Option          `json:"options,omitempty"`
	SubQuestions []QuestionContent `json:"sub_questions,omitempty"`
}

type Resource struct {
	Type        string `json:"type"`
	URL         string `json:"url,omitempty"`
	Alt         string `json:"alt,omitempty"`
	ImagePrompt string `json:"image_prompt,omitempty"`
	TTSText     string `json:"tts_text,omitempty"`
}

type Option struct {
	ID        string    `json:"id"`
	Text      string    `json:"text,omitempty"`
	Resource  *Resource `json:"resource,omitempty"`
	IsCorrect *bool     `json:"is_correct,omitempty"`
}

type QuestionAnswer struct {
	Value        any            `json:"value"`
	CorrectValue any            `json:"correct_value,omitempty"`
	AnalysisMode string         `json:"analysis_mode"`
	Explanation  string         `json:"explanation,omitempty"`
	Rubrics      []Rubric       `json:"rubrics,omitempty"`
	Configs      map[string]any `json:"configs,omitempty"`
}

type Rubric struct {
	Dimension   string  `json:"dimension"`
	MaxScore    float64 `json:"max_score"`
	Description string  `json:"description,omitempty"`
}

type EvaluateAnswerRequest struct {
	SessionID        string          `json:"session_id"`
	QuestionID       string          `json:"question_id"`
	QuestionTypeCode string          `json:"question_type_code"`
	Content          QuestionContent `json:"content"`
	ExpectedAnswer   QuestionAnswer  `json:"expected_answer"`
	StudentAnswer    any             `json:"student_answer"`
	AudioURL         string          `json:"audio_url,omitempty"`
	SubmittedAt      time.Time       `json:"submitted_at,omitempty"`
}

type EvaluateAnswerResult struct {
	IsCorrect     bool           `json:"is_correct"`
	Score         float64        `json:"score"`
	FullScore     float64        `json:"full_score"`
	CorrectAnswer CorrectAnswer  `json:"correct_answer"`
	Analysis      AnswerAnalysis `json:"analysis"`
	Metadata      map[string]any `json:"metadata,omitempty"`
}

type CorrectAnswer struct {
	Type       string      `json:"type"`
	Value      any         `json:"value,omitempty"`
	Values     []any       `json:"values,omitempty"`
	Options    []Option    `json:"options,omitempty"`
	SubAnswers []SubAnswer `json:"sub_answers,omitempty"`
}

type SubAnswer struct {
	SubID     string `json:"sub_id"`
	IsCorrect bool   `json:"is_correct"`
	Value     any    `json:"value,omitempty"`
}

type AnswerAnalysis struct {
	CorrectAnswer any    `json:"correct_answer,omitempty"`
	Explanation   string `json:"explanation,omitempty"`
	Analysis      string `json:"analysis,omitempty"`
}

type GenerateReportRequest struct {
	SessionID     string                 `json:"session_id"`
	StudentID     string                 `json:"student_id"`
	Subject       string                 `json:"subject"`
	Grade         int                    `json:"grade"`
	PracticeType  string                 `json:"practice_type"`
	Answers       []EvaluateAnswerResult `json:"answers"`
	Mastery       map[string]any         `json:"mastery,omitempty"`
	TotalTimeSecs int                    `json:"total_time_secs"`
}

type PracticeReportDraft struct {
	OverallScore         float64        `json:"overall_score"`
	CurrentAbility       float64        `json:"current_ability"`
	Confidence           float64        `json:"confidence"`
	AbilityLevel         string         `json:"ability_level"`
	Percentile           int            `json:"percentile"`
	KnowledgeScores      map[string]any `json:"knowledge_scores"`
	QuestionDistribution map[string]any `json:"question_distribution"`
	AbilityBreakdown     map[string]any `json:"ability_breakdown"`
	LearningSpeed        float64        `json:"learning_speed"`
	Consistency          float64        `json:"consistency"`
	Strengths            []string       `json:"strengths"`
	Weaknesses           []string       `json:"weaknesses"`
	Recommendations      []string       `json:"recommendations"`
}
